src/Microsoft.ML.Core/Data/VBuffer.cs

```diff
@@ -16,32 +16,35 @@ namespace Microsoft.ML.Runtime.Data
     /// </summary>
     public readonly struct VBuffer<T>
     {
-        /// <summary>
-        /// The logical length of the buffer.
-        /// </summary>
-        public readonly int Length;
+        private readonly T[] _values;
+        private readonly int[] _indices;
 
         /// <summary>
         /// The number of items explicitly represented. This is == Length when the representation
         /// is dense and &lt; Length when sparse.
         /// </summary>
-        public readonly int Count;
+        private readonly int _count;
+
+        /// <summary>
+        /// The logical length of the buffer.
+        /// </summary>
+        public readonly int Length;
 
         /// <summary>
         /// The values. Only the first Count of these are valid.
         /// </summary>
-        public readonly T[] Values;
+        public T[] Values => _values;
```

Comment:
>Values [](start = 19, length = 6)

Hmmm. I guess you are planning on removing this in subsequent PRs, but maybe we could hit this with an `ObsoleteArgument` so that until we weed out the stragglers still using this, if there are only truly a handful? Just an idea, if you're planning on doing this soonish so there's little chance of people continuing to use the old bad structure perhaps there's no need.

Reply:
It's still a decent amount of stragglers unfortunately. Currently the count is at over 200 usages.  :(

---
In reply to: [232433350](https://github.com/dotnet/machinelearning/pull/1580#discussion_r232433350) [](ancestors = 232433350)

Reply:
Adding an `Obsolete` attribute causes 200 build errors (since build warnings are turned to errors). So I can't do this. It will be removed "soonish".

---
In reply to: [232822938](https://github.com/dotnet/machinelearning/pull/1580#discussion_r232822938) [](ancestors = 232822938,232433350)